### DISCUSSION TITLE
what semantic kernel uses to store vector data and search vector data?

### DISCUSSION THREAD

**Discussion by xiguazhiPrince** [microsoft/semantic-kernel#104] [Q&A] [answered]
Hi I am a Java developer and am very confused about what semantic kernel uses to store vector data and search vector data.
 I can't understand the C++code, hahaha.
 Previously, I used redissearch, a vector search engine, to store and search in Java

**Comment by timlaverty** [comment, marked as answer]
Hi @xiguazhiPrince - 

We have a sample in PR right now (a GitHub repository exploration / Q&A chat sample) that showcases how to use memories; see: https://github.com/microsoft/semantic-kernel/pull/96 if you'd like to peruse.  See the GitHubskill.cs class for how we're wrapping memories/embeddings.  Then see QASkill/MemoryQuery, skprompt.txt for how we're pulling the memories back out of storage to include in questions (prompts) given to the AI.  This will go to Main shortly (lots of fast moving PR comments on it rn).

Memories (embedding) storage is abstracted and pluggable in Semantic Kernel.  In Main right now we enable storing them in a Dictionary (for quick testing) and SQLite on disk.  In the future we'll extend to Qdrant and other memory providers (in the cloud, local, etc.).

@awharrison-28  and @tawalke  are experts in this space and can go deeper if you'd like.

**Comment by kinthaiofficial** [comment]
Semantic Kernel's vector storage story has evolved quite a bit. Here's a quick breakdown:

**Built-in connectors** (as of recent SK versions):
- Azure Cognitive Search
- Qdrant
- Chroma
- Pinecone
- Weaviate
- Redis (via `RedisMemoryStore`)
- Postgres + pgvector (via `PostgresMemoryStore`)

```csharp
// Wiring up a vector store
var builder = Kernel.CreateBuilder();
builder.AddOpenAIChatCompletion("gpt-4o", apiKey);

// Option 1: In-memory (dev/testing)
builder.Services.AddSingleton<IMemoryStore, VolatileMemoryStore>();

// Option 2: Qdrant for production
builder.Services.AddQdrantMemoryStore("http://localhost:6333", 1536);

var kernel = builder.Build();
var memory = new SemanticTextMemory(
    kernel.Services.GetRequiredService<IMemoryStore>(),
    kernel.Services.GetRequiredService<ITextEmbeddingGenerationService>()
);
```

**Recommendation by use case**:
- Dev/testing: `VolatileMemoryStore` (in-memory, no setup)
- Production small-medium: pgvector (you likely already have Postgres)
- Production large-scale with filtering: Qdrant or Weaviate

For agent memory specifically (storing per-agent conversation history and learned facts), I've found that the metadata filtering support matters a lot — not all connectors implement `GetNearestMatchesAsync` with filter conditions equally well.

More on memory architecture for persistent agents: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture
